crypto/async/arch/async_posix.c

```diff
@@ -14,9 +14,27 @@
 
 # include <stddef.h>
 # include <unistd.h>
+# include <openssl/err.h>
+# include "internal/thread_once.h"
 
 #define STACKSIZE       32768
 
+static void *async_stack_alloc(size_t *num);
+static void async_stack_free(void *addr);
+
+static CRYPTO_RWLOCK *async_mem_lock;
+static CRYPTO_ONCE async_mem_init = CRYPTO_ONCE_STATIC_INIT;
+
+DEFINE_RUN_ONCE_STATIC(do_async_mem_init)
+{
+    async_mem_lock = CRYPTO_THREAD_lock_new();
```

Comment:
This lock needs to be freed on OpenSSL exit. See `async_deinit` which could be a hook to do this:

https://github.com/openssl/openssl/blob/a3b4cd1d6e307c1b1607ee9270658b5baf0b9870/crypto/async/async.c#L333-L350

A thought - this lock could be created as part of `async_init` instead and we just ensure that `OPENSSL_init_crypto(OPENSSL_INIT_ASYNC, NULL)` is called in `ASYNC_set_mem_functions` instead of the `RUN_ONCE` call. That might be more consistent with how the rest of async is initialised.

Reply:
Agreed.  I've added `int async_local_init(void)` and `void async_local_free(void)`, and for the POSIX variant init/deinit the mutex in those functions.